### ISSUE TITLE
stubs broken

### ISSUE THREAD

**Issue by Baaaaam** [cyclus/cyclus#1279]
I think my last PR #1278, to fix the ABI issues, broke the stubs archetypes. While compiling my own ([cyCLASS](https://github.com/Baaaaam/cyCLASS)) or the one of @mbmcgarry ([mbmore](https://github.com/mbmcgarry/mbmore)), we got:

> /Users/mouginot/.local/include/cyclus/prog_translator.h:6:10: fatal error: 'CoinPackedMatrix.hpp' file not found
> #include "CoinPackedMatrix.hpp"

Any one has an idea to fix those without breaking the ABI again ?

**Comment by scopatz** [comment]
Hi @Baaaaam, can you please post somewhere the full output.  Also, it seems that this is an API issue potentially, because it is failing to include a header.  An ABI issue would arise if it were to fail to link.

**Comment by Baaaaam** [comment]
there is the full output:
https://gist.github.com/Baaaaam/8712af4ead1941a69cef78d7e98eb6c0

**Comment by Baaaaam** [comment]
but In the change of @rwcarlsen `prog_translator.h` was not included any more in `toolkit/building_manager.h` as well as the include of `CoinPackedMatrix.hpp` (coming from `prog_translator.h`)

**Comment by scopatz** [comment]
That does not appear to be a clean build.  Please post the output from a clean build.  My guess is that the error you are seeing is from nit having a clean build.

**Comment by Baaaaam** [comment]
I think it was:
`python install.py -j4 --clean-build &> cyCLASS_install.txt`

**Comment by scopatz** [comment]
Also, try removing `-j4` as it makes it hard to reason about the output.

**Comment by Baaaaam** [comment]
> rm -rf build
> python install.py --clean-build &> cyCLASS_install.txt

https://gist.github.com/2f5c7a23dc38d3d39ec8c93561e2887f.git

**Comment by scopatz** [comment]
Do you actually have a file called `CoinPackedMatrix.hpp` on your system? Have you copied over all of the `cmake/` files and install.py from cyclus to your project?  It actually looks like your CMake configuration isn't trying to find Coin at all, which is why this is failing. 

The API change originally was probably done so that downstream packages wouldn't have to have the Coin headers installed.  An alternative change would revert what we have just reverted, remove the typedef and replace it with another struct that maybe throws an error if it is constructed.

However, I think that Coin is going to be a first class citizen on the same level as HDF5 or SQLite or Boost, we should probably just require that folks building archetypes have the Coin headers available.

**Comment by Baaaaam** [comment]
I agree, as you need Coin, HDF5 SQLite and Boost for Cyclus, we can update `stubs` to require them as well:
if you don't have them you can't compile Cyclus, so there is no point to compile you own Cyclus archetype!

**Comment by Baaaaam** [comment]
@scopatz: my folder was previously generated from `cycstub` and work well since the last PR.
I'll try to do what you suggest.

**Comment by scopatz** [comment]
Yeah, cycstub is probably also out of date with respect to this issue.

**Comment by Baaaaam** [comment]
I was able to fix it adding:

> # find coin and link to it
> FIND_PACKAGE( COIN REQUIRED )
> SET(STUB_INCLUDE_DIRS ${STUB_INCLUDE_DIRS} ${COIN_INCLUDE_DIR})
> set(LIBS ${LIBS} ${COIN_LIBRARIES})

in the `CMakeLists.txt` file

**Comment by Baaaaam** [comment]
Could we update the FIndCyclus.cmake to add those dependencies ?

**Comment by scopatz** [comment]
Yeah, we could / should.

**Comment by scopatz** [comment]
That is probably a change to be made in the cyclus repo itself and then copied to cycamore, etc.

**Comment by Baaaaam** [comment]
Do we want to do it for all the CYCLUS dependencies ?
 And use the install.py to recover it from CYCLUS folder ? So we don't have to propagate this change into cycamore repos any more ?

**Comment by scopatz** [comment]
I think trying to not propagate things and being too clever about the build system is what got us into this mess.  It is probably better to just have one place where the changes are made (cyclus repo) and then exact copies everywhere else.  

CMake is supposed to have some tools for installing the Find*.cmake files to a known path and then using those.  This was experimental for a long time, and might not be anymore, but trying to get that to work is ancillary to getting a release out.

**Comment by gonuke** [comment]
I think we have just discovered the reason for @rwcarlsen making this change in the first place.  I wondered why it was part of a PR on header leakage, and this is it.  By having a nested class/struct for `ProgTranslator::Context` it requires `prog_transalator.h` to be in `building_manager.h` which is in cycstub and all derivatives.

Cycstub & derivatives shouldn't need to know about COIN headers explicitly, so it's not a failure in the Find*.cmake.

I think we should revert this most recent revert, but it does result in an ABI change.  We should carefully look at how "public" that ABI change truly is.

**Comment by gonuke** [comment]
I want to pull out this point on Find*.cmake so it didn't get lost in my last post and since it is a main topic of this thread: I don't think CycStub or derivatives should have to know where to find COIN.  We don't expose any COIN API directly through the Cyclus API.  Cyclus does need COIN to build, but archetype developers don't/shouldn't need COIN headers to build!

**Comment by scopatz** [comment]
Hi @gonuke, sure maybe down stream packages shouldn't have to know about Coin headers, but requiring that they don't know about coin is asking for something to break.  And, the API & ABI did break.  

I believe that the most robust solution would be to keep the original API (even if it is replaced with a stub that throws an error if used) and have the core packages (stubs, cycamore, etc) also FindCoin.  I think this has the highest likelihood of not breaking downstream code.

**Comment by gonuke** [comment]
Architecturally, the DRE has been designed with a layer between the archetypes and COIN.  This is the purpose of the `ProgTranslator` in the first place (thoughts from @gidden are welcome here, too).  [1]

I belive an archetype developer should be able to develop an archetype without having COIN headers on their system.  Let's say that they install the Cyclus kernel in some way that doesn't require installing COIN headers (e.g. from a Debian package that depends on `coinor` and not `coinor-dev` [2]).  Does that mean that `cycstub`, etc, should fail?

Keeping the original API/ABI (and having it throw an error) in addition to this change is difficult because this change included a typedef that actually preserves as much of the API as possible.   In fact, because of this, the API did not change and only the ABI changed.

(magically) The ABI no longer includes `ProgTranslator::Context::Context()` but the API does through the typedef.  Therefore, I don't think we can legally declare a `ProgTranslator::Context::Context()`.

I tried an ugly polymorphism in which we declare `ProgTranslator::Context` as an empty derived class from `ProgTranslatorContext` rather than a typedef, but that doesn't work since  you can't change polymorphically (is that a word?) between the header - which uses the public base class to avoid header leakage - and the implementation - which uses the derived class to preserve the API.

[1] Does this imply that `ProgTranslator` should not be part of the public API of Cyclus?  (Recognizing that this would technically be a massive change to the API) Would this be accomplished by not installing the header file at all?  Note: no Cyclus headers or Cycamore files of any kind currently include `prog_translator.h` which is probably part of this design decision.

[2] the Debian package is currently setup to depend on the `-dev` version in order to make it easier to support both Ubuntu 14 & 16, but it is not strictly necessary.

**Comment by scopatz** [comment]
> Therefore, I don't think we can legally declare a ProgTranslator::Context::Context()

I don't think that I follow this.  Why wouldn't you be able to declare a constructor?

**Comment by gonuke** [comment]
Because `ProgTranslator::Context` is only a typedef to `ProgTranslatorContext`.  The constructor exists already and its `ProgTranslatorContext::ProgTranslatorContext`

**Comment by scopatz** [comment]
Sure, but `ProgTranslator::Context` and `ProgTranslatorContext` don't need to be the same type, so `ProgTranslator::Context` is free to define its own constructor - presumably one that raises a deprecation error.

**Comment by gonuke** [comment]
But if we don't use this typedef, then we really are changing the API as there are many places that continue to refer to `ProgTranslator::Context` as a valid type.  It is still perfectly valid to write new code in implementation files that refer to `ProgTranslator::Context` as those files will include `prog_translator.h` and be able to resolve the nested class.

Per the purpose of making this change, any new header file that wants to refer to an object of this type should refer to the actual type `ProgTranslatorContext` because it is not nested in `ProgTranslator` and can therefore be forward declared without including `prog_translator.h`.

**Comment by scopatz** [comment]
Maybe I am mistaken but there only seem to be three instances outside of the this class that refer to `ProgTranslator::Context`, and they are all in `toolkit/building_manager.cc`

```console
scopatz@artemis ~ $ cd cyclus/src/
scopatz@artemis ~/cyclus/src v1.4.0-rc2 $ grep -nr ProgTranslator::Context
prog_translator.h:85:  ProgTranslator::Context ctx_;
toolkit/building_manager.cc:23:    ProgTranslator::Context ctx;
toolkit/building_manager.cc:33:                             ProgTranslator::Context& ctx,
toolkit/building_manager.cc:67:                             ProgTranslator::Context& ctx,
scopatz@artemis ~/cyclus/src v1.4.0-rc2 $ cd ../tests/
scopatz@artemis ~/cyclus/tests v1.4.0-rc2 $ grep -nr ProgTranslator::Context
```

Since they are all in an implementation file in function bodies, doesn't this make it effectively private?  There are no instances of its usage in the tests nor in cycamore (as expected).  

I think we could change these few instances of `ProgTranslator::Context` to `ProgTranslatorContext` without causing any ABI breakages going forward.  

Also, the typedef does break APIs because you can no longer call `ProgTranslator::Context::Context()` or `ProgTranslator::Context::~Context()` explicitly.   Instead, we have `ProgTranslator::Context::ProgTranslatorContext()` and `ProgTranslator::Context::~ProgTranslatorContext()`.  Not this would ever be used like this, but the ABI breakage occurs because of this API break.

**Comment by gonuke** [comment]
Right you are on API breakage.  I was noting that implicit construction of a `ProgTranslator::Context` object was still valid, but hadn't thought of the explicit constructor.... <sigh>

**Comment by gonuke** [comment]
Given our philosophy on API/ABI stability, I think we have to do this, as painful as I find it to leave this poorly designed API/ABI in place for the 1e-10 probability that it is being invoked somewhere.... `<sighs again>`

**Comment by scopatz** [comment]
Yeah, sigh is basically how I feel.  And this is a prime thing to fix up on the next major release.  APIs do slowly decay until the next time the slate can be wiped clean :)

**Comment by gonuke** [comment]
I'll give it a try now...

**Comment by gonuke** [comment]
Thanks for working this through with me @scopatz

**Comment by scopatz** [comment]
Thanks for caring about stability, @gonuke, unlike some OR packages I could name :).

**Comment by Baaaaam** [comment]
I think this have been fix by PR #1281.

closing.
